### PR TITLE
[Draft] Support for CSV via JDBC

### PR DESCRIPTION
**Pull request by hellonico**
This is a POC for a CSV via JDBC driver for metabase.
The csv.clj file is based on a mix of the h2 and the mysql driver and some custom connection query string code. 

The CSV/JDBC driver used is the one below:
http://csvjdbc.sourceforge.net/

Metabase is started with the below:

```
java -cp csvjdbc-1.0-34.jar:sources/target/uberjar/metabase.jar metabase.core
```

The hello.csv file is in the CSV folder. Each CSV file is a table.

And the results gives: 
csv_4
csv_3
csv_2
csv_0
csv_1

###### Before submitting the PR, please make sure you do the following 
-  [x] If there are changes to the backend codebase, run the backend tests with `lein test && lein eastwood && lein bikeshed && lein docstring-checker && ./bin/reflection-linter`
-  [x] Run the frontend and integration tests with  `yarn && yarn run prettier && yarn run lint && yarn run flow && ./bin/build version uberjar && yarn run test`)
-  [x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
(unless it's a tiny documentation change).

### CODE REVIEW

File: src/metabase/db/spec.clj [hunk 1]
```diff
@@ -14,6 +14,17 @@
           :subname db}
          (dissoc opts :db)))
 
+(defn csv
+  "Create a database specification for a csv database. Opts should include a key
+  for :db which is the path to the database file."
+  [{:keys [db]
+    :or {db "h2.db"}
```
Comment (camsaul) [inline src/metabase/db/spec.clj]: Why would your CSV file be named `h2.db?

File: src/metabase/driver/csv.clj [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+  clojure.lang.Named
+  (getName [_] "CSV"))
+
+(u/strict-extend CSVDriver
+  driver/IDriver
+  (merge (sql/IDriverSQLDefaultsMixin)
+         {:date-interval                     (u/drop-first-arg date-interval)
+          :details-fields                    (constantly [{:name         "db"
+                                                           :display-name "Connection String"
+                                                           :placeholder  "file:/Users/camsaul/bird_sightings/toucans"
```
Comment (camsaul) [inline src/metabase/driver/csv.clj]: excellent placeholder 💯

Comment (hellonico) [reply src/metabase/driver/csv.clj]: Should work on your machine ;) 
Added a better one.

File: src/metabase/db/spec.clj (hunk 1 above)
Comment (hellonico) [reply src/metabase/db/spec.clj]: -> data.csv

File: src/metabase/driver/csv.clj (hunk 2 above)
Comment (camsaul) [reply src/metabase/driver/csv.clj]: Oh I wasn’t being sarcastic I genuinely liked it

**Comment by hellonico** [comment]
I just added support for more options, which allows to use the quite cool feature of using split csv files.

**Comment by hellonico** [comment]
With the below set of CSV files:
[CONCATS.zip](https://github.com/metabase/metabase/files/2003150/CONCATS.zip)

And the following configuration:
setup

We can execute simple select queries:
simple select

And group by queries on the set of files.
group by

Note that the group is added based on the file pattern matching.

**Comment by hellonico** [comment]
Added support (for CSV data sources) for field filters, as well as proper linked filter on dashboards.

question
dashboard

**Comment by jbrooksuk** [comment]
When will we see this in Metabase? :)

**Comment by hellonico** [comment]
@jbrooksuk I can't wait either. Although at this stage, my private branch is ahead of public branch. 
Was waiting for instruction about where / what to fix to get this accepted.

**Comment by jbrooksuk** [comment]
I wonder if @tlrobinson can help out? :)

**Comment by tvvignesh** [comment]
Merge this someone please. Thanks.

**Comment by gaokeji52** [comment]
I follow all the steps but I just can't get the database source with CSV, anyone helps?

I downloaded and copied the two files in the same path under java folder in Windows;

Using cmd to startup metabase like this
java -cp csvjdbc-1.0-34.jar;metabase.jar metabase.core

**Comment by hellonico** [comment]
@gaokeji52 
I sync-ed the branch with the latest metabase master, and compiled again.
Looks like the it's still working fine and the CSV datasource shows up as expected.

screen shot 2018-10-04 at 9 39 20 am

**Comment by gaokeji52** [comment]
@hellonico 
Hi, I tried to follow the steps to build a developer version of metabase with your merge branch, however, I failed at the last step when installing the leiningen in Windows7, 

Is it possible that I can load the csvjdbc with this command only? In windows7,

java -cp csvjdbc-1.0-34.jar;metabase.jar metabase.core

article here:
https://github.com/metabase/metabase/blob/master/docs/developers-guide.md#build-metabase

**Comment by hellonico** [comment]
Hi @gaokeji52 

I've created a sample projects:
https://github.com/hellonico/metabase-csv-sample
And indication on downloading the jar files from my maven repository, so you do not need to rebuild the jars. 

Give it a try !

**Comment by gaokeji52** [comment]
@hellonico 

Woo, that's pretty cool! That's really kind of you!

I tried to compile the jars for the whole day, but still not able to work it out,

image

**Comment by hellonico** [comment]
@gaokeji52

Ahh ..that's my fault. That missing dependency is not on a public maven repository.
I'll fix that, but in the meantime, please use the instructions I sent earlier on.

**Comment by hellonico** [comment]
Yes, the link to the metabase.jar file was wrong.

Please 
[download](http://repository.hellonico.info/repository/hellonico/org/hellonico/metabase/metabase-02-83d1a40581400d46e58c6c528afc030496e408d0/metabase-metabase-02-83d1a40581400d46e58c6c528afc030496e408d0.jar) it again.

**Comment by gaokeji52** [comment]
@hellonico 

Thanks, man, it works!

**Comment by rNianios** [comment]
This seems like a great feature. Any chance to be included in a released version of Metabase?

**Comment by jbrooksuk** [comment]
Is there any chance we'd see this feature implemented within the next few versions?

**Comment by gscottstukey** [comment]
<3 Bump

**Comment by tlrobinson** [comment]
Hey all,

Thank you for this PR and apologies for the long delay responding. We've recently done some work to support drivers as plugins in Metabase and as of the next release (v0.32) we will be asking driver developers to first publish new drivers in their own repositories (typically named `metabase-*-driver`). If the driver fully passes the driver tests we will link to it from our documentation, and if a driver gains significant usage we will consider merging it into Metabase itself.

Here is some documentation about publishing a driver as a plugin in a separate repository: https://github.com/metabase/metabase/wiki/Writing-a-Driver:-Packaging-a-Driver-&-Metabase-Plugin-Basics#drivers-shipped-as-3rd-party-plugins

Here are two example drivers: https://github.com/metabase/sudoku-driver https://github.com/metabase/crate-driver

We still have some work to do on the documentation. You can follow progress and provide feedback on that here: https://github.com/metabase/metabase/issues/9348

Thank you.

**Comment by jbrooksuk** [comment]
@tlrobinson how does this work with Docker installs?

**Comment by tlrobinson** [comment]
@jbrooksuk I think the easiest way right now is to build you own Docker image based on one of ours, e.x.:

```
FROM metabase/metabase-head:latest

ADD https://github.com/metabase/sudoku-driver/releases/download/1.0.0/sudoku.metabase-driver.jar /app/plugins/
```

(`metabase/metabase-head` is the latest `master`, once we release `v0.32` you'd do `FROM metabase/metabase:v0.32.0`)

Save to something like `Dockerfile-sudoku` then do `docker build -f Dockerfile-sudoku -t metabase-with-sudoku .` then do whatever you want with the `metabase-with-sudoku` tagged image (run it locally, push it to your own registry, etc)

**Comment by jbrooksuk** [comment]
Oh! That's easy enough, thanks!

**Pull request closed without merging**